### ISSUE TITLE
--sync-timeout flag not respected

### ISSUE THREAD

**Issue by ionfury** [fluxcd/flux#3372]
The `--sync-timeout` flag is recommended to prevent context deadline exceeded errors when executing long running generator commands: https://docs.fluxcd.io/en/1.21.0/references/fluxyaml-config-files/

**Describe the bug**

The `--sync-timeout` flag is not being respected when a value is set.  The command will time out after 60s, regardless of the value passed in via `--sync-timeout`.

**To Reproduce**


1. Clone the repository at https://github.com/ionfury/flux-test
2. Install the flux manifests on a cluster with `kubectl apply -f flux/manifests.yaml`
3. Get the ssh key with the command `fluxctl identity --k8s-fwd-ns flux` and add it to your repositories deploy keys.

**Expected behavior**

The manifests at `manifests/spin` to be applied successfully.

The `manfiests/spin` directory has a custom kustomize generator, `SpinGenerator`, which takes a time value as a parameter and waits the given time before completing.  This represents some long-running kustomize process.

When the `SpinGenerator` is given a value less than the value passed to `--sync-timeout` the manifests should be applied successfully.  By default, the repository is configured with a timeout of 2m, and a spin of 70s.  This causes a failure with the message `component=sync-loop err="loading resources from repo: error executing generator command \"XDG_CONFIG_HOME=$(pwd) kustomize build --enable_alpha_plugins .\" from file \"../.flux.yaml\": context deadline exceeded\nerror output:\n\ngenerated output:\n"`

In my testing, I found that the operation would time out in 60s regardless of the value passed in the `--sync-timeout` parameter.  For example, a timeout of 20s and a spin of 30s would complete successfully.

I found the same results testing flux versions 1.21.0 through 1.17.0

All testing was done on a local cluster created with kind.

**Logs**

```
❯ kubectl logs deployment/flux -n flux                                                                                                                                                                                                                           ─╯
I1124 22:43:08.836512       6 main.go:242] Kubernetes client verbosity level set to 5
ts=2020-11-24T22:43:08.8365835Z caller=main.go:259 version=1.21.0
ts=2020-11-24T22:43:08.8366458Z caller=main.go:412 msg="using kube config: \"/root/.kube/config\" to connect to the cluster"
I1124 22:43:08.837326       6 merged_client_builder.go:122] Using in-cluster configuration
I1124 22:43:08.838486       6 reflector.go:150] Starting reflector *v1beta1.CustomResourceDefinition (0s) from pkg/mod/k8s.io/client-go@v0.17.11/tools/cache/reflector.go:105
I1124 22:43:08.838520       6 reflector.go:185] Listing and watching *v1beta1.CustomResourceDefinition from pkg/mod/k8s.io/client-go@v0.17.11/tools/cache/reflector.go:105
ts=2020-11-24T22:43:08.8611775Z caller=main.go:492 component=cluster identity=/etc/fluxd/ssh/identity
ts=2020-11-24T22:43:08.8612326Z caller=main.go:493 component=cluster identity.pub="<snip>"
ts=2020-11-24T22:43:08.8612638Z caller=main.go:498 host=https://10.96.0.1:443 version=kubernetes-v1.16.3
ts=2020-11-24T22:43:08.8613223Z caller=main.go:510 kubectl=/usr/local/bin/kubectl
ts=2020-11-24T22:43:08.8624366Z caller=main.go:527 ping=true
ts=2020-11-24T22:43:08.8628722Z caller=main.go:566 warning="--docker-config not used; pre-flight check failed" err="open /docker/config.json: no such file or directory"
ts=2020-11-24T22:43:08.8647638Z caller=main.go:666 url=ssh://git@github.com/ionfury/flux-test user="\"Weave Flux\"" email="\"tnowak@noreply.github.com\"" signing-key= verify-signatures-mode=none sync-tag=flux-sync state=git readonly=false registry-disable-scanning=false notes-ref=flux set-author=false git-secret=false sops=false
ts=2020-11-24T22:43:08.8648208Z caller=main.go:772 upstream="no upstream URL given"
ts=2020-11-24T22:43:08.8648951Z caller=main.go:803 metrics-addr=:9121
ts=2020-11-24T22:43:08.8650262Z caller=sync.go:51 component=daemon warning="failed to load last-synced resources. sync event may be inaccurate" err="git repo not ready: git repo has not been cloned yet"
ts=2020-11-24T22:43:08.8650967Z caller=loop.go:108 component=sync-loop err="git repo not ready: git repo has not been cloned yet"
ts=2020-11-24T22:43:08.8651222Z caller=images.go:17 component=sync-loop msg="polling for new images for automated workloads"
ts=2020-11-24T22:43:08.8651582Z caller=images.go:27 component=sync-loop msg="no automated workloads"
ts=2020-11-24T22:43:08.8689317Z caller=main.go:795 addr=:3030
ts=2020-11-24T22:43:18.6090082Z caller=loop.go:134 component=sync-loop event=refreshed url=ssh://git@github.com/ionfury/flux-test branch=master HEAD=08048aebefca542891fbf61bf48fe811684482fd
ts=2020-11-24T22:43:48.6687615Z caller=sync.go:61 component=daemon info="trying to sync git changes to the cluster" old=30c7eda175b2fbb45acb5a5b4e0d381c508d0c29 new=08048aebefca542891fbf61bf48fe811684482fd
ts=2020-11-24T22:44:58.6319054Z caller=loop.go:108 component=sync-loop err="loading resources from repo: error executing generator command \"XDG_CONFIG_HOME=$(pwd) kustomize build --enable_alpha_plugins .\" from file \"../.flux.yaml\": context deadline exceeded\nerror output:\n\ngenerated output:\n"
```

**Additional context**

- Flux version: 1.21.0 - 1.17.0
- Kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-19T13:57:45Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-12-04T07:23:47Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
- Git provider: github.com
- Container registry provider: docker.io

**Comment by kingdonb** [comment]
Is this possibly addressed by #3381, I wonder?

(haven't tested yet, leaving breadcrumbs.) @ionfury If you are able to verify that #3381 resolves the issue, I will prioritize merging it for the next release. It looks like there is a gap in the functions covered by the sync timeout value.

**Comment by kingdonb** [comment]
Bump - I'd like to be sure it gets resolved in #3442 for 1.22.0, if there are still folks suffering from this.

**Comment by kingdonb** [comment]
I will give reproducing this issue a go for myself before the 1.22.0 release, and see if #3381 can resolve it.

Would appreciate if anyone with better visibility into this issue than I have currently could drop me a note, here or privately in DM, so I can be sure I've understood correctly. The fix proposed in #3381 looks very simple but just obscure enough to make me wonder if I've really understood the issue.

If it can't be resolved inside of the next day or two, then it will be bumped and can be included in 1.22.1 later instead.

**Comment by kingdonb** [comment]
This is resolved in #3381
